src/libraries/System.Text.Json/src/System/Text/Json/Serialization/JsonConverterOfT.cs

```diff
@@ -310,33 +331,59 @@ internal override sealed bool TryReadAsObject(ref Utf8JsonReader reader, JsonSer
         }
 
         /// <summary>
-        /// Overridden by the nullable converter to prevent boxing of values by the JIT.
+        /// Performance optimization. Overridden by the nullable converter to prevent boxing of values by the JIT.
+        /// Although the JIT added support to detect this IL pattern (box+isinst+br), it still manifests in TryWrite().
+        /// </summary>
+        internal virtual bool IsNull(in T value) => value is null;
+
+        /// <summary>
+        /// Performance optimization to determine if IsNull() above should be called.
         /// </summary>
-        internal virtual bool IsNull(in T value) => value == null;
+        /// <returns></returns>
+        internal virtual bool IsNullableOfT() => false;
 
+#if NET6_0_OR_GREATER
+        [MethodImpl(MethodImplOptions.AggressiveOptimization)]
+#endif
         internal bool TryWrite(Utf8JsonWriter writer, in T value, JsonSerializerOptions options, ref WriteStack state)
         {
             if (writer.CurrentDepth >= options.EffectiveMaxDepth)
             {
                 ThrowHelper.ThrowJsonException_SerializerCycleDetected(options.EffectiveMaxDepth);
             }
 
-            if (default(T) is null && !HandleNullOnWrite && IsNull(value))
+            // We do not pass null values to converters unless HandleNullOnWrite is true. Null values for properties were
+            // already handled in GetMemberAndWriteJson() so we don't need to check for IgnoreNullValues here.
+            if (_autoWriteNulls)
             {
-                // We do not pass null values to converters unless HandleNullOnWrite is true. Null values for properties were
-                // already handled in GetMemberAndWriteJson() so we don't need to check for IgnoreNullValues here.
-                writer.WriteNullValue();
-                return true;
+                // Optimized check to see if 'IsNull()' or 'is null' should be used.
+                if (_isNullableOfT)
```

Comment:
The JIT can eliminate this branch if we change this to
```suggestion
                if (typeof(T).IsValueType)
```

cf. https://sharplab.io/#v2:C4LghgzgtgPgAgJgIwFgBQcAMACAygCzACcAHAGTACMA6AJQFcA7YASygFMBuddOAZmyJsAYQA2kCNgDe6bHOyz5AbQBSLYAHF2jdkRYBjABTAAniXYB7AGaGWzAJT2Auormr1WnXqOnz128wA/I4uaPKCAnBIAGzYdsDYAGIWFgA8ACoAfAHY6dgAbmCi9Oz2CmHyMhXhcixW2IYAJuxWYPSiwIbpZSySjO2i9q41VTVjcfXGZpY23dQAkhAAakUl6dND1eNyo9vjcADs2JjcW9sAvsPbV/tHSKcXN3JPgkcID/KXaOdAA=